nearcore/src/cold_storage.rs

```diff
@@ -117,8 +135,66 @@ fn cold_store_copy_result_to_string(result: &anyhow::Result<ColdStoreCopyResult>
     }
 }
 
+#[derive(Debug)]
+enum ColdStoreInitialMigrationResult {
+    /// Cold storage was already initialized
+    NoNeedForMigration,
+    /// Performed a successful cold storage migration
+    SuccessfulMigration,
+    /// Migration was interrupted by keep_going flag
+    MigrationInterrupted,
+}
+
+/// This function performes intial population of cold storage if needed.
+/// Migration can be interrupted via `keep_going` flag.
+///
+/// First, checks that hot store is of kind `Archive`. If not, no migration needed.
+/// Then, captures hot final head BEFORE the migration, as migration is performed during normal neard run.
+/// If hot final head is not set, returns Err.
+/// Otherwise:
+/// 1. performed migration
+/// 2. updates head to saved hot final head
+///
+/// Any Ok status means that this function should not be retried:
+/// - either migration was performed (now or earlier)
+/// - or migration was interrupted, which means the `keep_going` flag was set to `false`
+///   which means that everything cold store thread related has to stop
+///
+/// Error status means that for some reason migration cannot be performed.
+fn cold_store_initial_migration(
+    keep_going: Arc<AtomicBool>,
+    hot_store: &Store,
+    cold_db: Arc<ColdDB>,
+) -> anyhow::Result<ColdStoreInitialMigrationResult> {
+    // We only need to perform the migration if hot store is of kind Archive and cold store doesn't have a head yet
+    if hot_store.get_db_kind()? != Some(near_store::metadata::DbKind::Archive)
+        || hot_store.get(DBCol::BlockMisc, COLD_HEAD_KEY)?.is_some()
+    {
+        return Ok(ColdStoreInitialMigrationResult::NoNeedForMigration);
+    }
+
+    tracing::info!(target: "cold_store", "Starting initial population of cold store");
+
+    // If FINAL_HEAD is not set for hot storage something isn't right and we will probably fail in `update_cold_head`.
+    // Let's fail early.
+    let hot_final_head = hot_store
+        .get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?
+        .ok_or_else(|| anyhow::anyhow!("FINAL_HEAD not found in hot storage"))?;
+    let hot_final_head_height = hot_final_head.height;
+
+    // TODO take batch_size from config
+    if copy_all_data_to_cold(cold_db.clone(), &hot_store, 500_000_000, keep_going)? {
+        tracing::info!(target: "cold_store", "Initial population was successful, writing cold head and hot db kind");
```

Comment:
> writing cold head

agreed
> and hot db kind

I don't think this is the case and it shouldn't be. The prepare-hot command is the one that will write the hot db kind and it will only do so in the rpc backup db. 

Can you add the new cold head height to the log line too?